lib/libmuser.c

```diff
@@ -1105,39 +1088,88 @@ free_sparse_mmap_areas(lm_reg_info_t *reg_info)
 }
 
 static int
-lm_caps_init(lm_ctx_t *lm_ctx, lm_cap_t *caps, int nr_caps)
+pci_config_setup(lm_ctx_t *lm_ctx, const lm_dev_info_t *dev_info)
 {
+    lm_reg_info_t *cfg_reg;
+    lm_reg_info_t zero_reg;
+    int i;
     int err;
 
-    assert(lm_ctx);
-    assert(caps);
+    assert(lm_ctx != NULL);
+    assert(dev_info != NULL);
 
-    if (!nr_caps) {
-        return 0;
+    // Convenience pointer.
+    cfg_reg = &lm_ctx->pci_info.reg_info[LM_DEV_CFG_REG_IDX];
+
+    // Set a default config region if none provided.
+    memset(&zero_reg, 0, sizeof(zero_reg));
+    if (memcmp(cfg_reg, &zero_reg, sizeof(*cfg_reg)) == 0) {
+        cfg_reg->flags = LM_REG_FLAG_RW;
+        cfg_reg->size = PCI_CFG_SPACE_SIZE;
+    } else {
+        // Validate the config region provided.
+        if ((cfg_reg->flags != LM_REG_FLAG_RW) ||
+            ((cfg_reg->size != PCI_CFG_SPACE_SIZE) &&
+             (cfg_reg->size != PCI_CFG_SPACE_EXP_SIZE))) {
+            return EINVAL;
```

Comment:
A debug message explaining why it failed would be nice. Then the comment could go away.

Reply:
I agree. But I think we should do a separate review of our debug/log messages and make it all consistent there. Right now, it's a bit unclear when we log what. If that's ok I'll not add this now, but instead do it separately across the entire codebase.